src/shared/sleep-config.c

```diff
@@ -136,6 +136,53 @@ int read_battery_capacity_percentage(void) {
         return battery_capacity;
 }
 
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *dirpath = NULL;
+        int stored_discharge_rate, r;
+        char *dis_rate;
+        sd_id128_t machine_id = SD_ID128_NULL;
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get machine-id: %m");
+
+        dirpath = path_join("/var/lib/systemd/sleep", SD_ID128_TO_STRING(machine_id), "battery_discharge_percentage_rate_per_hour");
+        if (!dirpath)
+               return log_oom_debug();
```

Comment:
Code to determine the file path is duplicated between the two functions, let's pull this out into a helper function.

Reply:
Only this part is left. Except that fixed all bugs. Will take this tomorrow.